src/symbolic.jl

```diff
@@ -502,36 +457,38 @@ is the result of trying to do the sub and `did_change` is a bool specifying
 whether or not any substitutions happened. This is used to know when to break
 out of a recursion.
 =#
+
 function _subs(s::Symbol, d::Associative, a...)
     haskey(d, s) && return (d[s], true)
-
-    # also check for canonical form (s, 0)
-    haskey(d, (s, 0)) && return (d[(s, 0)], true)
-
     (s, false)
 end
 
 _subs(x::Number, d::Associative, a...) = (x, false)
 
-function _subs(ex::Expr, d::Associative, funcs::Set{Symbol})
+function _subs(ex::Expr, d::Associative, funcs::Set{Symbol}, shiftit=false)
+
     if is_time_shift(ex)
+
         var, shift = arg_name_time(ex)
-        if haskey(d, (var, shift))
+        if (haskey(d, (var, shift)))
             new_ex = d[(var, shift)]
             return new_ex, true
-        end
-
-        if haskey(d, var)
-            new_ex = time_shift(var, shift, funcs, d)
+        elseif haskey(d, ex)
+            # I hassume the hash of ex is easy to compute
+            new_ex = d[ex]
             return new_ex, true
         end
-
-        # repeat for (var, 0)
-        if haskey(d, (var, 0))
-            new_ex = time_shift(d[(var, 0)], shift, funcs, d)
-            return new_ex, true
+        if shiftit
+            for ts in (-1,1)
+                if (haskey(d, (var, shift+ts)))
+                    new_ex = time_shift(d[(var, shift+ts)], -ts)
```

Comment:
Should this be `ts` or `-ts` when doing the time shift?

Reply:
This while block shouldn't be there, if we don't do any time shift in subs.